control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This must be included before any of our internal headers because ir/std.h
+// places a vector class in the global namespace, which breaks protobuf.
+#include "p4/config/p4info.pb.h"
+
+#include <boost/optional.hpp>
+#include <boost/variant.hpp>
+#include <google/protobuf/util/json_util.h>
+#include <iostream>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/coreLibrary.h"
+#include "frontends/p4/fromv1.0/v1model.h"
+#include "frontends/p4/methodInstance.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+#include "lib/ordered_set.h"
+#include "PI/pi_base.h"
+
+#include "piSerializer.h"
+
+namespace P4 {
+namespace ControlPlaneAPI {
+
+// XXX(seth): Here are the known issues:
+// - Direct counters are not yet supported.
+// - Direct meters are not yet supported.
+// - Action profiles are not yet supported.
+// - The @id pragma is not supported for header instances and especially not for
+//   header stack instances, which currently requires allowing the programmer to
+//   provide a sequence of ids in the @id pragma. Note that @id *is* supported
+//   for header instance *fields*; this is just about the instances themselves.
+// - Locals are intentionally not exposed, but this prevents table match keys
```

Comment:
In P4-14 header instance is a very clear concept. So it is in BMv2: they are global variables.
In P4-16 you can have instances of a header type in multiple places. They can be fields in the struct with type H in v1model.p4, or they can be local variables, or they can even be fields of a struct which is used to instantiate local variables anywhere in a local scope.